### PR TITLE
Make UserIpReader configuration more flexible.

### PR DESCRIPTION
**Pull request by demiankatz**
This is a follow-up to #1681.

TODO
- [x] Add documentation to security wiki page.

### CODE REVIEW

**Comment by demiankatz** [comment]
As noted over at #1681, this is my attempt to improve the flexibility of the IP forwarding support. Feedback would be appreciated if @lahmann or others have time to comment.

**Comment by lahmann** [comment]
Thanks @demiankatz! I think this is as good as it can be - this leaves enough options for configuration of specific needs but covers the most common cases.
Though I would change the default `HTTP_X_FORWARDED_FOR` behaviour to `first` as the standard behaviour should be that the clients IP is listed first (cf. https://en.wikipedia.org/wiki/X-Forwarded-For#Format which is referring to the apparent original specification of the custom header in squid).

**Comment by demiankatz** [comment]
@lahmann, I set the `HTTP_X_FORWARDED_FOR` behavior to `last` in the example because the `first` value is more likely to be spoofed. At least, in my experimentation, if I set a fake X-Forwarded-For header using a browser plugin, my proxy appends my real IP after the spoofed one before passing the X-Forwarded-For header along to VuFind. Thus, the "last" value is guaranteed to be real, but the "first" value might be fake. I'm sure this can vary from device to device and depending on configuration, of course. If you think "first" would be a better real-world example, I'm certainly happy to change the example in config.ini (since it is just that: an example).

Perhaps we should also expand upon all of this in the wiki, and include a link to the wiki from the config.ini comments. I'll add a TODO and make sure I write up some richer documentation before merging this.

**Comment by lahmann** [comment]
This makes sense, although in order to address such a scenario appropriately it would be necessary to be able to configure the load balancer/proxy IP or include the proxy configuration for VuFind in order to take the left IP address next to the last known load balancer/proxy IP.
But then again I agree with your comment in config.ini that one could extent `VuFind\Net\UserIpReader`... so, I am undecided to which extent this should cover any known scenario.

I agree that this is worth an additional Wiki-page.

**Comment by demiankatz** [comment]
Thanks again, @lahmann. I have added notes on the [security wiki page](https://vufind.org/wiki/administration:security#proxies_and_ip_authentication). I'm going to merge this now in the interest of further improving the current bleeding-edge code, but again, I'm happy to continue this conversation. None of this should be considered "set in stone" until we decide to schedule the 7.0.1 release. If we need to add or change the configuration format in the meantime, I'm open to suggestions.

File: module/VuFind/src/VuFind/Net/UserIpReader.php [hunk 1]
```diff
@@ -48,18 +48,19 @@ class UserIpReader
...
      */
     protected $allowForwardedIps;
 
     /**
      * Constructor
      *
      * @param Parameters $server            Server parameters
-     * @param bool       $allowForwardedIps Should we respect the X-Forwarded-For
-     * header?
+     * @param bool       $allowForwardedIps Forwarded header configuration string
```
Comment (berndoberknapp) [inline module/VuFind/src/VuFind/Net/UserIpReader.php:62]: This probably should be @param string|bool.

Comment (demiankatz) [reply module/VuFind/src/VuFind/Net/UserIpReader.php:62]: Thanks for the catch; fixed here: https://github.com/vufind-org/vufind/commit/9202ef95dd28be398630d262c79219cf8ef8b0a3

**Pull request merged**